discretize_model oversample non-integer ranges fails

<!-- This comments are hidden when you submit the issue,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- Please have a search on our GitHub repository to see if a similar
issue has already been posted.
If a similar issue is closed, have a quick look to see if you are satisfied
by the resolution.
If not please go ahead and open an issue! -->

<!-- Please check that the development version still produces the same bug.
You can install development version with
pip install git+https://github.com/astropy/astropy
command. -->

### Description
<!-- Provide a general description of the bug. -->

`astropy.convolution.utils.discretize_model` with arg `mode=oversample` works with `x_range`, `y_range` ranges that have non-integer lower and upper limits as long as the oversample factor ensures the number of steps between the lower and upper are integer.

This no longer works as designed due to a recent PR that changed the way the oversampling grid is done from `np.arange()` to `np.linspace()`: https://github.com/astropy/astropy/pull/9293

### Expected behavior
<!-- What did you expect to happen. -->

According to https://github.com/astropy/astropy/issues/5472#issuecomment-264557888 non-integer ranges should work, as long as the difference between upper and lower limit are integer.

### Actual behavior
<!-- What actually happened. -->
<!-- Was the output confusing or poorly described? -->

The code raises an exception because `num` arg to `np.linspace` is not of type `int`, e.g.:

```python
gunagala/psf.py:204: in pixellated
    return discretize_model(self, xrange, yrange, mode='oversample', factor=10)
../../miniconda3/envs/gunagala/lib/python3.8/site-packages/astropy/convolution/utils.py:188: in discretize_model
    return discretize_oversample_2D(model, x_range, y_range, factor)
../../miniconda3/envs/gunagala/lib/python3.8/site-packages/astropy/convolution/utils.py:266: in discretize_oversample_2D
    x = np.linspace(x_range[0] - 0.5 * (1 - 1 / factor),
<__array_function__ internals>:5: in linspace
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

start = -10.45, stop = 10.45, num = 210.0, endpoint = True, retstep = False, dtype = None, axis = 0

    @array_function_dispatch(_linspace_dispatcher)
    def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
                 axis=0):
        """
        [docstring clipped]

        """
>       num = operator.index(num)
E       TypeError: 'float' object cannot be interpreted as an integer
```


### Steps to Reproduce
<!-- Ideally a code example could be provided so we can run it ourselves. -->
<!-- If you are pasting code, use triple backticks (```) around
your code snippet. -->
<!-- If necessary, sanitize your screen output to be pasted so you do not
reveal secrets like tokens and passwords. -->

Just run the following to get the error:

```python
from astropy.modeling.functional_models import Gaussian2D
from astropy.convolution import discretize_model
gmodel = Gaussian2D(amplitude=1.0, x_mean=5.,
                                      y_mean=125., x_stddev=0.75, y_stddev=3)
xrange = (-10.5, 10.5)
yrange = (-10.5, 10.5)
discretize_model(gmodel, xrange, yrange, mode='oversample', factor=10)
```

Note this works just fine cause the ranges have `int` values

```python
from astropy.modeling.functional_models import Gaussian2D
from astropy.convolution import discretize_model
gmodel = Gaussian2D(amplitude=1.0, x_mean=5.,
                                      y_mean=125., x_stddev=0.75, y_stddev=3)
xrange = (-10, 10)
yrange = (-10, 10)
discretize_model(gmodel, xrange, yrange, mode='oversample', factor=10)
```

### System Details
<!-- Even if you do not think this is necessary, it is useful information for the maintainers.
Please run the following snippet and paste the output below:
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import numpy; print("Numpy", numpy.__version__)
import astropy; print("astropy", astropy.__version__)
import scipy; print("Scipy", scipy.__version__)
import matplotlib; print("Matplotlib", matplotlib.__version__)
-->

macOS-10.15.6-x86_64-i386-64bit
Python 3.8.5 (default, Aug  5 2020, 03:39:04)
[Clang 10.0.0 ]
Numpy 1.19.1
astropy 4.0.1.post1
Scipy 1.5.2
Matplotlib 3.3.1

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -929,6 +929,7 @@ astropy.convolution
 - Fixed a bug where a float-typed integers in the argument ``x_range`` of
   ``astropy.convolution.utils.discretize_oversample_1D`` (and the 2D version as
   well) fails because it uses ``numpy.linspace``, which requires an ``int``.
+  [#10696]
 
 astropy.coordinates
 ^^^^^^^^^^^^^^^^^^^

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -926,6 +926,10 @@ astropy.constants
 astropy.convolution
 ^^^^^^^^^^^^^^^^^^^
 
+- Fixed a bug where a float-typed integers in the argument ``x_range`` of
+  ``astropy.convolution.utils.discretize_oversample_1D`` (and the 2D version as
+  well) fails because it uses ``numpy.linspace``, which requires an ``int``.
+
 astropy.coordinates
 ^^^^^^^^^^^^^^^^^^^
 

--- a/astropy/convolution/tests/test_discretize.py
+++ b/astropy/convolution/tests/test_discretize.py
@@ -183,7 +183,7 @@ def test_float_x_range_exception():
         return x ** 2 + y ** 2
     with pytest.raises(ValueError) as exc:
         discretize_model(f, (-10.002, 11.23))
-    assert exc.value.args[0] == ("The difference between the upper an lower"
+    assert exc.value.args[0] == ("The difference between the upper and lower"
                                  " limit of 'x_range' must be a whole number.")
 
 
@@ -192,7 +192,7 @@ def test_float_y_range_exception():
         return x ** 2 + y ** 2
     with pytest.raises(ValueError) as exc:
         discretize_model(f, (-10, 11), (-10.002, 11.23))
-    assert exc.value.args[0] == ("The difference between the upper an lower"
+    assert exc.value.args[0] == ("The difference between the upper and lower"
                                  " limit of 'y_range' must be a whole number.")
 
 

--- a/astropy/convolution/utils.py
+++ b/astropy/convolution/utils.py
@@ -249,7 +249,7 @@ def discretize_oversample_1D(model, x_range, factor=10):
     # Evaluate model on oversampled grid
     x = np.linspace(x_range[0] - 0.5 * (1 - 1 / factor),
                     x_range[1] - 0.5 * (1 + 1 / factor),
-                    num = (x_range[1] - x_range[0]) * factor)
+                    num=int((x_range[1] - x_range[0]) * factor))
 
     values = model(x)
 
@@ -265,10 +265,10 @@ def discretize_oversample_2D(model, x_range, y_range, factor=10):
     # Evaluate model on oversampled grid
     x = np.linspace(x_range[0] - 0.5 * (1 - 1 / factor),
                     x_range[1] - 0.5 * (1 + 1 / factor),
-                    num = (x_range[1] - x_range[0]) * factor)
+                    num=int((x_range[1] - x_range[0]) * factor))
     y = np.linspace(y_range[0] - 0.5 * (1 - 1 / factor),
                     y_range[1] - 0.5 * (1 + 1 / factor),
-                    num = (y_range[1] - y_range[0]) * factor)
+                    num=int((y_range[1] - y_range[0]) * factor))
 
     x_grid, y_grid = np.meshgrid(x, y)
     values = model(x_grid, y_grid)

--- a/astropy/convolution/utils.py
+++ b/astropy/convolution/utils.py
@@ -159,12 +159,12 @@ def discretize_model(model, x_range, y_range=None, mode='center', factor=10):
         raise ValueError('discretize_model only supports 1-d and 2-d models.')
 
     if not float(np.diff(x_range)).is_integer():
-        raise ValueError("The difference between the upper an lower limit of"
+        raise ValueError("The difference between the upper and lower limit of"
                          " 'x_range' must be a whole number.")
 
     if y_range:
         if not float(np.diff(y_range)).is_integer():
-            raise ValueError("The difference between the upper an lower limit of"
+            raise ValueError("The difference between the upper and lower limit of"
                              " 'y_range' must be a whole number.")
 
     if ndim == 2 and y_range is None:

